In [1]:
from ultralytics import YOLO
from ultralytics import settings


In [ ]:
#settings.reset()
# View all settings
print(settings)

In [15]:
#this worked and should be the basis for the small dataset, the project variable forces YOLO to this root

from ultralytics import YOLO

model = YOLO("yolo11n-seg.pt")

#pathdataset = "E:/Techlabs_AI/Digital wardrop/datasets/YOLOsmall/YOLOsmall.yaml"
pathdataset = "E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOsmall/YOLOsmall.yaml"


# Train the model using the 'coco8.yaml' dataset for 3 epochs
results = model.train(data=pathdataset, epochs=3, project="E:/Techlabs_AI/digital_wardrobe/eWardrobe/YOLO/runs")

# Evaluate the model's performance on the validation set
results = model.val()

# Perform object detection on an image using the model
results = model("E:/Techlabs_AI/Digital wardrop/datasets/small/test/000011.jpg")

Ultralytics 8.3.184  Python-3.12.10 torch-2.8.0+cpu CPU (12th Gen Intel Core(TM) i9-12900KF)
engine\trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOsmall/YOLOsmall.yaml, degrees=0.0, deterministic=True, device=cpu, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=3, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolo11n-seg.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=train, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patie

train: Scanning E:\Techlabs_AI\digital_wardrobe\eWardrobe\datasets\YOLOsmall\train\labels.cache... 100 images, 0 backgrounds, 0 corrupt: 100%|██████████| 100/100 [00:00<?, ?it/s]

val: Fast image access  (ping: 0.00.0 ms, read: 131.830.1 MB/s, size: 31.0 KB)



e:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\venv312\Lib\site-packages\torch\utils\data\dataloader.py:666: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  warnings.warn(warn_msg)
val: Scanning E:\Techlabs_AI\digital_wardrobe\eWardrobe\datasets\YOLOsmall\val\labels.cache... 100 images, 0 backgrounds, 0 corrupt: 100%|██████████| 100/100 [00:00<?, ?it/s]

Plotting labels to E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\train\labels.jpg... 



e:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\venv312\Lib\site-packages\torch\utils\data\dataloader.py:666: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  warnings.warn(warn_msg)


optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000588, momentum=0.9) with parameter groups 90 weight(decay=0.0), 101 weight(decay=0.0005), 100 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\train
Starting training for 3 epochs...

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


        1/3         0G      1.369      5.708      4.177      1.643         21        640: 100%|██████████| 7/7 [00:13<00:00,  1.93s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.06it/s]

                   all        100        136       0.02       0.02     0.0276     0.0248       0.02       0.02     0.0276    0.00552



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


        2/3         0G      1.158      4.783      4.135      1.488         12        640: 100%|██████████| 7/7 [00:12<00:00,  1.78s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.08it/s]

                   all        100        136     0.0391      0.129     0.0328     0.0204     0.0391      0.129     0.0354      0.017



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


        3/3         0G      1.077      3.999      4.079      1.387         17        640: 100%|██████████| 7/7 [00:12<00:00,  1.81s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.02it/s]

                   all        100        136      0.019      0.307     0.0295      0.016     0.0165      0.242      0.028     0.0119



3 epochs completed in 0.014 hours.
Optimizer stripped from E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\train\weights\last.pt, 6.0MB
Optimizer stripped from E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\train\weights\best.pt, 6.0MB

Validating E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\train\weights\best.pt...
Ultralytics 8.3.184  Python-3.12.10 torch-2.8.0+cpu CPU (12th Gen Intel Core(TM) i9-12900KF)
YOLO11n-seg summary (fused): 113 layers, 2,837,103 parameters, 0 gradients, 9.6 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.13it/s]


                   all        100        136     0.0636      0.157     0.0459     0.0249     0.0636      0.157     0.0485     0.0234
   short_sleeved_shirt         31         31        0.2     0.0323      0.105      0.063        0.2     0.0323      0.105      0.063
    long_sleeved_shirt         17         17     0.0667      0.118     0.0468     0.0355     0.0667      0.118     0.0468      0.019
                  vest          7          7          0          0          0          0          0          0          0          0
                shorts          4          4    0.00309        0.5    0.00392    0.00157    0.00309        0.5    0.00392    0.00136
              trousers         12         12        0.1     0.0833     0.0882     0.0793        0.1     0.0833     0.0882     0.0529
                 skirt         29         29          0          0          0          0          0          0          0          0
   short_sleeved_dress         25         25       0.25       0.04   

val: Scanning E:\Techlabs_AI\digital_wardrobe\eWardrobe\datasets\YOLOsmall\val\labels.cache... 100 images, 0 backgrounds, 0 corrupt: 100%|██████████| 100/100 [00:00<?, ?it/s]
e:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\venv312\Lib\site-packages\torch\utils\data\dataloader.py:666: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  warnings.warn(warn_msg)
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:03<00:00,  2.14it/s]


                   all        100        136     0.0478      0.146     0.0365     0.0198     0.0478      0.146     0.0393     0.0195
   short_sleeved_shirt         31         31      0.167     0.0323     0.0872     0.0523      0.167     0.0323     0.0872     0.0523
    long_sleeved_shirt         17         17          0          0          0          0          0          0          0          0
                  vest          7          7          0          0          0          0          0          0          0          0
                shorts          4          4    0.00319        0.5    0.00397    0.00159    0.00319        0.5    0.00397    0.00137
              trousers         12         12     0.0909     0.0833      0.084     0.0756     0.0909     0.0833      0.084     0.0504
                 skirt         29         29          0          0          0          0          0          0          0          0
   short_sleeved_dress         25         25        0.2       0.04   

In [ ]:
#small model test

from ultralytics import YOLO

model = YOLO("yolo11n-seg.yaml")

pathdataset = "E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOsmall/YOLOsmall.yaml"

# Train the model for 3 epochs
results = model.train(data=pathdataset, epochs=3)

# Evaluate the model's performance on the validation set
results = model.val()

# Perform object detection on an image using the model
results = model("E:/Techlabs_AI/Digital wardrop/datasets/small/test/000011.jpg")

The training settings for YOLO models encompass various hyperparameters and configurations used during the training process. These settings influence the model's performance, speed, and accuracy. Key training settings include batch size, learning rate, momentum, and weight decay. Additionally, the choice of optimizer, loss function, and training dataset composition can impact the training process. Careful tuning and experimentation with these settings are crucial for optimizing performance.

| Argument       | Type              | Default | Description |
|----------------|------------------|---------|-------------|
| model          | str              | None    | Path to a `.pt` pretrained model or `.yaml` config file. Defines model structure or initializes weights. |
| data           | str              | None    | Path to dataset config (e.g., `coco8.yaml`). Defines train/val data, class names, etc. |
| epochs         | int              | 100     | Total number of epochs (full passes through dataset). |
| time           | float            | None    | Max training time (hours). Overrides `epochs` if set. |
| patience       | int              | 100     | Early stopping patience (epochs without improvement). |
| batch          | int/float        | 16      | Batch size. Supports `int`, auto (`-1`), or fraction (e.g., `0.7`). |
| imgsz          | int              | 640     | Target image size. Resizes to square images. |
| save           | bool             | True    | Save training checkpoints and final weights. |
| save_period    | int              | -1      | Save model checkpoints every N epochs. `-1` disables. |
| cache          | bool             | False   | Cache dataset images (`ram`, `disk`, or `False`). |
| device         | int/str/list     | None    | Training device(s): `0`, `[0,1]`, `cpu`, `mps`, `-1`. |
| workers        | int              | 8       | Data loading workers per GPU rank. |
| project        | str              | None    | Project directory for training outputs. |
| name           | str              | None    | Name of run (subdir inside project). |
| exist_ok       | bool             | False   | Allow overwriting of an existing `project/name`. |
| pretrained     | bool/str         | True    | Use pretrained model (bool or path to weights). |
| optimizer      | str              | auto    | Optimizer (`SGD`, `Adam`, `AdamW`, etc.). |
| seed           | int              | 0       | Random seed for reproducibility. |
| deterministic  | bool             | True    | Force deterministic algorithms (slower, reproducible). |
| single_cls     | bool             | False   | Train all classes as one (binary classification). |
| classes        | list[int]        | None    | Train only on specific class IDs. |
| rect           | bool             | False   | Use rectangular batches (minimal padding). |
| multi_scale    | bool             | False   | Train with variable image sizes. |
| cos_lr         | bool             | False   | Cosine learning rate scheduler. |
| close_mosaic   | int              | 10      | Disable mosaic augmentation for last N epochs. |
| resume         | bool             | False   | Resume from last checkpoint. |
| amp            | bool             | True    | Use Automatic Mixed Precision (faster, less memory). |
| fraction       | float            | 1.0     | Fraction of dataset to use. |
| profile        | bool             | False   | Profile ONNX/TensorRT speeds. |
| freeze         | int/list         | None    | Freeze first N layers or list of layers. |
| lr0            | float            | 0.01    | Initial learning rate. |
| lrf            | float            | 0.01    | Final learning rate fraction. |
| momentum       | float            | 0.937   | Momentum (SGD) / beta1 (Adam). |
| weight_decay   | float            | 0.0005  | L2 regularization term. |
| warmup_epochs  | float            | 3.0     | Warmup epochs for LR. |
| warmup_momentum| float            | 0.8     | Warmup momentum start value. |
| warmup_bias_lr | float            | 0.1     | Warmup bias learning rate. |
| box            | float            | 7.5     | Weight for bounding box loss. |
| cls            | float            | 0.5     | Weight for classification loss. |
| dfl            | float            | 1.5     | Weight for distribution focal loss. |
| pose           | float            | 12.0    | Weight for pose loss. |
| kobj           | float            | 2.0     | Weight for keypoint objectness loss. |
| nbs            | int              | 64      | Nominal batch size (loss normalization). |
| overlap_mask   | bool             | True    | Merge object masks or keep separate. |
| mask_ratio     | int              | 4       | Downsample ratio for segmentation masks. |
| dropout        | float            | 0.0     | Dropout rate (classification). |
| val            | bool             | True    | Enable validation during training. |
| plots          | bool             | False   | Save plots of metrics and predictions. |


In [ ]:
#Small Experimentation center

model = YOLO("yolo11n-seg.yaml") #train from scratch
#model = YOLO("yolo11n-seg.pt") # pretrained weights

data="E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOfull/YOLOfull.yaml"
#data="E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOsmall/YOLOsmall.yaml"

pretrained="False"
project="E:/Techlabs_AI/digital_wardrobe/eWardrobe/YOLO/runs"
name="Run_full_yaml_pretrainFalse"
resume="False"
fraction="1.0" #float
epochs=10
batch= -1




SyntaxError: invalid syntax (2585166842.py, line 9)

In [1]:
import torch
print(torch.__version__)          # 2.7.1+cu128
print(torch.version.cuda)         # 12.8
print(torch.cuda.is_available())  # should be True
print(torch.cuda.get_device_name(0))  # RTX 4080

2.8.0+cu128
12.8
True
NVIDIA GeForce RTX 4080


In [ ]:
# I used this to keep on training from the last.pt file

from ultralytics import YOLO


#model_cfg="yolo11s-seg.pt"   # can be .yaml (architecture) or .pt (weights+arch)
model_cfg= "E:/Techlabs_AI/digital_wardrobe/eWardrobe/YOLO/runs/Run_full_pt_pretrainTrue_continue6/weights/last.pt"
data="E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOfull/YOLOfull.yaml"
#data="E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOsmall/YOLOsmall.yaml",
pretrained=True,               # bool or str (path to custom checkpoint)
project="E:/Techlabs_AI/digital_wardrobe/eWardrobe/YOLO/runs"
name="Run_full_pt_pretrainTrue_continue7"
resume=False
fraction=1
epochs=20
batch=16
imgsz=640
plots=True

"""
Experimentation center for YOLO training.
"""
# Initialize model
model = YOLO(model_cfg)

# Train
results = model.train(
    data=data,
    pretrained=pretrained,
    project=project,
    name=name,
    resume=resume,
    fraction=fraction,
    epochs=epochs,
    batch=batch,
    plots=plots,
    imgsz=imgsz
)

    


New https://pypi.org/project/ultralytics/8.3.192 available  Update with 'pip install -U ultralytics'
Ultralytics 8.3.184  Python-3.12.10 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 4080, 16376MiB)
engine\trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOfull/YOLOfull.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=20, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=E:/Techlabs_AI/digital_wardrobe/eWardrobe/YOLO/runs/Run_ful

train: Scanning E:\Techlabs_AI\digital_wardrobe\eWardrobe\datasets\YOLOfull\train\labels.cache... 191961 images, 0 backgrounds, 1 corrupt: 100%|██████████| 191961/191961 [00:00<?, ?it/s]

train: E:\Techlabs_AI\digital_wardrobe\eWardrobe\datasets\YOLOfull\train\images\164121.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [1.01068]


val: Fast image access  (ping: 0.00.0 ms, read: 132.938.4 MB/s, size: 49.4 KB)


val: Scanning E:\Techlabs_AI\digital_wardrobe\eWardrobe\datasets\YOLOfull\val\labels.cache... 32153 images, 0 backgrounds, 0 corrupt: 100%|██████████| 32153/32153 [00:00<?, ?it/s]


Plotting labels to E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\Run_full_pt_pretrainTrue_continue7\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 90 weight(decay=0.0), 101 weight(decay=0.0005), 100 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\Run_full_pt_pretrainTrue_continue7
Starting training for 20 epochs...

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       1/20      2.81G     0.6043      2.132       0.85      1.005         34        640: 100%|██████████| 11998/11998 [33:55<00:00,  5.90it/s] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:29<00:00,  6.74it/s]


                   all      32153      52490      0.773      0.763      0.813      0.717      0.625      0.609      0.631      0.416

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       2/20      4.52G     0.6213      2.166     0.8713      1.012         29        640: 100%|██████████| 11998/11998 [20:18<00:00,  9.85it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:27<00:00,  6.80it/s]


                   all      32153      52490      0.773      0.755      0.815      0.718      0.625      0.604      0.632      0.416

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       3/20      4.53G     0.6756      2.314     0.9705      1.042         26        640: 100%|██████████| 11998/11998 [19:40<00:00, 10.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.61it/s]


                   all      32153      52490      0.781      0.743       0.81      0.713       0.63      0.595      0.627      0.411

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       4/20      4.53G      0.711      2.422      1.041      1.063         39        640: 100%|██████████| 11998/11998 [20:18<00:00,  9.84it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.59it/s]


                   all      32153      52490       0.77      0.746      0.804      0.707      0.622      0.594      0.623      0.406

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       5/20      4.53G     0.6964      2.392      1.016      1.055         36        640: 100%|██████████| 11998/11998 [20:04<00:00,  9.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:33<00:00,  6.57it/s]


                   all      32153      52490      0.768      0.741      0.801      0.703      0.616       0.59      0.619      0.404

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       6/20      4.53G     0.6852      2.366     0.9983      1.049         33        640: 100%|██████████| 11998/11998 [20:00<00:00,  9.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:33<00:00,  6.56it/s]


                   all      32153      52490      0.772      0.736        0.8      0.702      0.618      0.587      0.617      0.404

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       7/20      4.53G     0.6763      2.335     0.9805      1.044         27        640: 100%|██████████| 11998/11998 [20:03<00:00,  9.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.58it/s]


                   all      32153      52490      0.775      0.737        0.8      0.702      0.623      0.586      0.618      0.404

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       8/20      4.53G     0.6698      2.316     0.9677       1.04         28        640: 100%|██████████| 11998/11998 [20:05<00:00,  9.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:31<00:00,  6.64it/s]


                   all      32153      52490      0.775      0.738      0.802      0.704      0.624      0.586      0.619      0.405

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


       9/20      4.53G     0.6615      2.293     0.9514      1.035         32        640: 100%|██████████| 11998/11998 [19:59<00:00, 10.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:35<00:00,  6.46it/s]


                   all      32153      52490      0.774      0.743      0.804      0.705      0.625      0.588      0.621      0.406

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      10/20      4.53G     0.6547      2.279     0.9416      1.032         33        640: 100%|██████████| 11998/11998 [20:05<00:00,  9.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.58it/s]


                   all      32153      52490      0.761      0.754      0.805      0.707      0.612      0.601      0.622      0.407
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      11/20      4.53G     0.5868      1.962     0.6836      1.015         15        640: 100%|██████████| 11998/11998 [19:54<00:00, 10.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.61it/s]


                   all      32153      52490      0.769      0.752      0.808      0.709      0.619      0.599      0.625      0.409

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      12/20      4.53G     0.5742      1.931      0.663      1.006         11        640: 100%|██████████| 11998/11998 [19:56<00:00, 10.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.59it/s]


                   all      32153      52490      0.775      0.749      0.809      0.711      0.624      0.598      0.626       0.41

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      13/20      4.53G     0.5647      1.901     0.6451     0.9987         15        640: 100%|██████████| 11998/11998 [19:47<00:00, 10.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.59it/s]


                   all      32153      52490      0.767       0.76      0.811      0.713      0.616      0.609      0.628      0.412

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      14/20      4.53G     0.5538      1.871     0.6282     0.9919         12        640: 100%|██████████| 11998/11998 [19:49<00:00, 10.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:31<00:00,  6.62it/s]


                   all      32153      52490      0.772      0.759      0.813      0.717      0.623      0.606      0.631      0.415

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      15/20      4.54G      0.544      1.839     0.6101     0.9843         13        640: 100%|██████████| 11998/11998 [19:47<00:00, 10.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:30<00:00,  6.67it/s]


                   all      32153      52490      0.771      0.762      0.816       0.72       0.62       0.61      0.632      0.417

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      16/20      4.54G     0.5323      1.806     0.5899     0.9774         13        640: 100%|██████████| 11998/11998 [19:52<00:00, 10.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:31<00:00,  6.63it/s]


                   all      32153      52490       0.77       0.77      0.817      0.722       0.62      0.615      0.634      0.419

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      17/20      4.54G     0.5202       1.77     0.5685     0.9691         14        640: 100%|██████████| 11998/11998 [19:49<00:00, 10.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:33<00:00,  6.53it/s]


                   all      32153      52490      0.769      0.775       0.82      0.724      0.619       0.62      0.637      0.421

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      18/20      4.54G     0.5091      1.729     0.5464      0.962         13        640: 100%|██████████| 11998/11998 [19:48<00:00, 10.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.60it/s]


                   all      32153      52490       0.78      0.767      0.822      0.727      0.627      0.615      0.638      0.423

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      19/20      4.54G     0.4966       1.69     0.5225     0.9544         14        640: 100%|██████████| 11998/11998 [19:46<00:00, 10.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:34<00:00,  6.52it/s]


                   all      32153      52490      0.783       0.77      0.823      0.729      0.631      0.616       0.64      0.424

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


      20/20      4.54G     0.4845      1.643     0.4979     0.9465         12        640: 100%|██████████| 11998/11998 [19:52<00:00, 10.07it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:32<00:00,  6.61it/s]


                   all      32153      52490       0.79      0.765      0.825      0.731      0.637      0.613      0.641      0.426

20 epochs completed in 7.742 hours.
Optimizer stripped from E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\Run_full_pt_pretrainTrue_continue7\weights\last.pt, 6.0MB
Optimizer stripped from E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\Run_full_pt_pretrainTrue_continue7\weights\best.pt, 6.0MB

Validating E:\Techlabs_AI\digital_wardrobe\eWardrobe\YOLO\runs\Run_full_pt_pretrainTrue_continue7\weights\best.pt...
Ultralytics 8.3.184  Python-3.12.10 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 4080, 16376MiB)
YOLO11n-seg summary (fused): 113 layers, 2,837,103 parameters, 0 gradients, 9.6 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 1005/1005 [02:17<00:00,  7.33it/s]


                   all      32153      52490       0.79      0.765      0.825      0.731      0.636      0.613      0.641      0.426
   short_sleeved_shirt      12371      12556       0.89      0.911      0.958      0.888      0.883      0.905      0.948      0.584
    long_sleeved_shirt       5916       5966      0.797      0.845      0.892      0.806      0.777      0.828      0.859      0.433
 short_sleeved_outwear        142        142      0.637      0.346      0.467      0.396      0.598      0.325      0.383      0.165
  long_sleeved_outwear       1999       2011      0.764      0.883      0.872      0.798      0.664      0.767       0.68      0.236
                  vest       2098       2113      0.773      0.815      0.864      0.754      0.765      0.809      0.852      0.634
                 sling        317        322      0.784       0.54      0.646      0.556      0.778      0.537      0.629      0.471
                shorts       4133       4167      0.888      0.854   

YOLO11 can process different types of input sources for inference, as shown in the table below. The sources include static images, video streams, and various data formats. The table also indicates whether each source can be used in streaming mode with the argument stream=True ✅. Streaming mode is beneficial for processing videos or live streams as it creates a generator of results instead of loading all frames into memory.


| Source        | Example                                   | Type         | Notes                                                                 |
|---------------|-------------------------------------------|--------------|----------------------------------------------------------------------|
| image         | `'image.jpg'`                             | str or Path  | Single image file.                                                    |
| URL           | `'https://ultralytics.com/images/bus.jpg'`| str          | URL to an image.                                                      |
| screenshot    | `'screen'`                                | str          | Capture a screenshot.                                                 |
| PIL           | `Image.open('image.jpg')`                 | PIL.Image    | HWC format with RGB channels.                                         |
| OpenCV        | `cv2.imread('image.jpg')`                 | np.ndarray   | HWC format with BGR channels uint8 (0-255).                           |
| numpy         | `np.zeros((640,1280,3))`                  | np.ndarray   | HWC format with BGR channels uint8 (0-255).                           |
| torch         | `torch.zeros(16,3,320,640)`               | torch.Tensor | BCHW format with RGB channels float32 (0.0-1.0).                      |
| CSV           | `'sources.csv'`                           | str or Path  | CSV file containing paths to images, videos, or directories.          |
| video ✅       | `'video.mp4'`                             | str or Path  | Video file in formats like MP4, AVI, etc.                             |
| directory ✅   | `'path/'`                                 | str or Path  | Path to a directory containing images or videos.                      |
| glob ✅        | `'path/*.jpg'`                            | str          | Glob pattern to match multiple files. Use the `*` as a wildcard.      |
| YouTube ✅     | `'https://youtu.be/LNwODJXcvt4'`          | str          | URL to a YouTube video.                                               |
| stream ✅      | `'rtsp://example.com/media.mp4'`          | str          | URL for streaming protocols (RTSP, RTMP, TCP) or IP address.          |
| multi-stream ✅| `'list.streams'`                          | str or Path  | `.streams` text file with one stream URL per row (batch-size = count).|
| webcam ✅      | `0`                                       | int          | Index of the connected camera device to run inference on.             |



Inference arguments:

| Argument       | Type        | Default                  | Description                                                                                                                                                   |
|----------------|-------------|--------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------|
| source         | str         | `'ultralytics/assets'`   | Data source for inference: image path, video file, directory, URL, or device ID (e.g., webcam). Supports many formats.                                        |
| conf           | float       | 0.25                     | Minimum confidence threshold for detections. Detections below this value are discarded. Helps reduce false positives.                                          |
| iou            | float       | 0.7                      | IoU threshold for Non-Maximum Suppression (NMS). Lower values remove more overlapping boxes, useful to reduce duplicates.                                      |
| imgsz          | int/tuple   | 640                      | Image size for inference. Can be a single int (square resize) or a tuple `(h, w)`. Impacts speed vs. accuracy.                                                 |
| rect           | bool        | True                     | If enabled: minimally pads shorter side for stride compatibility. If disabled: pads to a square.                                                              |
| half           | bool        | False                    | Enables half-precision (FP16) inference. Speeds up inference on supported GPUs with minor accuracy tradeoff.                                                   |
| device         | str         | None                     | Compute device: `"cpu"`, `"cuda:0"`, `"0"` (first GPU), etc.                                                                                                   |
| batch          | int         | 1                        | Batch size for inference (only for directory, video, or `.txt` sources). Larger batch → higher throughput.                                                     |
| max_det        | int         | 300                      | Maximum number of detections per image. Prevents too many boxes in dense scenes.                                                                               |
| vid_stride     | int         | 1                        | Frame stride for video input. Higher values skip frames (faster but less temporal resolution).                                                                 |
| stream_buffer  | bool        | False                    | If False: drops old frames for real-time speed. If True: buffers all frames (higher latency).                                                                  |
| visualize      | bool        | False                    | Enables visualization of model features during inference. Helpful for debugging/model insights.                                                                |
| augment        | bool        | False                    | Enables test-time augmentation (TTA). Improves robustness but slows inference.                                                                                 |
| agnostic_nms   | bool        | False                    | Enables class-agnostic NMS (overlapping boxes merged across classes). Useful when class overlap occurs.                                                        |
| classes        | list[int]   | None                     | Filters detections to specific class IDs only. Good for focusing on relevant classes in multi-class tasks.                                                     |
| retina_masks   | bool        | False                    | Returns high-res segmentation masks (original image size). If False: masks are resized to inference size.                                                      |
| embed          | list[int]   | None                     | Extracts feature vectors/embeddings from specific layers. Useful for clustering, similarity search, etc.                                                       |
| project        | str         | None                     | Project directory name where prediction outputs are saved (if `save=True`).                                                                                   |
| name           | str         | None                     | Subdirectory name within project folder for saving outputs (if `save=True`).                                                                                   |
| stream         | bool        | False                    | Returns a generator of `Results` objects for long videos or many images. Saves memory.                                                                         |
| verbose        | bool        | True                     | Controls logging verbosity. If True: detailed inference logs shown in terminal.                                                                                |


Visualization arguments:

| Argument     | Type        | Default        | Description                                                                                                                                                                      |
|--------------|-------------|----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| show         | bool        | False          | If True, displays annotated images/videos in a window. Useful for immediate visual feedback during development or testing.                                                       |
| save         | bool        | False / True   | Saves annotated images/videos to file. Defaults to **True in CLI**, **False in Python**. Useful for documentation, analysis, or sharing.                                         |
| save_frames  | bool        | False          | When processing videos, saves individual frames as images. Helpful for extracting frames or frame-by-frame analysis.                                                             |
| save_txt     | bool        | False          | Saves detection results in text format: `[class] [x_center] [y_center] [width] [height] [confidence]`. Useful for integration with other tools.                                  |
| save_conf    | bool        | False          | Includes confidence scores in saved text files. Provides richer detail for post-processing and evaluation.                                                                       |
| save_crop    | bool        | False          | Saves cropped images of detected objects. Useful for dataset creation, augmentation, or focused analysis.                                                                        |
| show_labels  | bool        | True           | Displays labels for each detection in the visual output. Provides immediate understanding of detected objects.                                                                   |
| show_conf    | bool        | True           | Displays confidence score next to each label in the visual output. Helps gauge model certainty.                                                                                  |
| show_boxes   | bool        | True           | Draws bounding boxes around detected objects. Essential for visually locating objects in images/videos.                                                                          |
| line_width   | None or int | None           | Sets bounding box line width. If `None`, auto-adjusts based on image size. Allows customization for clarity (e.g., thick lines for presentations, thin for dense detections).     |


In [ ]:
#Ultralytics predict sample code

from ultralytics import YOLO

# Load a model
model = YOLO("yolo11n.pt")  # pretrained YOLO11n model

# Run batched inference on a list of images
results = model(["image1.jpg", "image2.jpg"], stream=True)  # return a generator of Results objects

# Process results generator
for result in results:
    boxes = result.boxes  # Boxes object for bounding box outputs
    masks = result.masks  # Masks object for segmentation masks outputs
    keypoints = result.keypoints  # Keypoints object for pose outputs
    probs = result.probs  # Probs object for classification outputs
    obb = result.obb  # Oriented boxes object for OBB outputs
    result.show()  # display to screen
    result.save(filename="result.jpg")  # save to disk

In [ ]:
from ultralytics import YOLO
# Inference 

# Load your trained model
model = YOLO(r"runs/Run_full_pt_pretrainTrue_continue7/weights/best.pt")

# List of test images
test_images = [
    #r"E:/Techlabs_AI/digital_wardrobe/eWardrobe/datasets/YOLOfull/test/images/000011.jpg",
    r"E:/Techlabs_AI/digital_wardrobe/test/5.jpeg"
]

# Run inference
results = model(test_images, stream=True)

# Process results
for i, result in enumerate(results):
    print(f"\n🔹 Results for image {i+1}:")
    print("Boxes:", result.boxes)   # bounding box info
    print("Masks:", result.masks)   # segmentation masks
    result.show()                   # show image with predictions
    result.save(filename=f"prediction_{i+1}.jpg")  # save annotated image



0: 480x640 1 short_sleeved_dress, 1 vest_dress, 45.0ms

🔹 Results for image 1:
Boxes: ultralytics.engine.results.Boxes object with attributes:

cls: tensor([11.,  9.], device='cuda:0')
conf: tensor([0.9044, 0.8369], device='cuda:0')
data: tensor([[4.1167e+02, 5.7438e+02, 8.6664e+02, 1.3534e+03, 9.0440e-01, 1.1000e+01],
        [1.2013e+03, 5.5796e+02, 1.7427e+03, 1.5355e+03, 8.3687e-01, 9.0000e+00]], device='cuda:0')
id: None
is_track: False
orig_shape: (1536, 2048)
shape: torch.Size([2, 6])
xywh: tensor([[ 639.1555,  963.8795,  454.9743,  778.9946],
        [1471.9586, 1046.7344,  541.4116,  977.5525]], device='cuda:0')
xywhn: tensor([[0.3121, 0.6275, 0.2222, 0.5072],
        [0.7187, 0.6815, 0.2644, 0.6364]], device='cuda:0')
xyxy: tensor([[ 411.6684,  574.3823,  866.6427, 1353.3768],
        [1201.2528,  557.9581, 1742.6644, 1535.5106]], device='cuda:0')
xyxyn: tensor([[0.2010, 0.3739, 0.4232, 0.8811],
        [0.5865, 0.3633, 0.8509, 0.9997]], device='cuda:0')
Masks: ultralytics.e

Estimation of training time for improved results:

| Model          | Img size | Time / epoch | Time for 100 epochs   |
| -------------- | -------- | ------------ | --------------------- |
| **n (nano)**   | 640      | \~22 min     | \~37 h                |
| **n (nano)**   | 896      | \~44 min     | \~73 h (\~3 days)     |
| **n (nano)**   | 1024     | \~55 min     | \~91 h (\~3.8 days)   |
| **s (small)**  | 640      | \~44 min     | \~73 h (\~3 days)     |
| **s (small)**  | 896      | \~88 min     | \~146 h (\~6 days)    |
| **s (small)**  | 1024     | \~110 min    | \~183 h (\~7.5 days)  |
| **m (medium)** | 640      | \~77 min     | \~128 h (\~5.3 days)  |
| **m (medium)** | 896      | \~154 min    | \~256 h (\~10.5 days) |
| **m (medium)** | 1024     | \~192 min    | \~320 h (\~13 days)   |
